### PR TITLE
Cw 262 better handle user exchange amount below minimum or maximum trade size

### PR DESCRIPTION
**Pull request by Blazebrain**
Issue Number (if Applicable): Fixes #

# Description
Better handle user exchange amount below minimum or maximum trade size.

- [x] Using the minimum and maximum limits for each exchange, when a user types in a value that is below the minimum input or above the maximum limit, an error text  comes up, informing the user that it is above or below the limit. 

- [x] The "Convert to" value and "Convert from" value is left blank when this occurs, instead of it populating with the above/below limit values. 

- [x] The "Convert to" value is left blank when it's a variable exchange, while the "Convert from" value is left blank if it's a fixed exchange

- [x] The error text comes under the "Convert from" text field when it's a variable exchange while it comes under the "Convert to" text field if it's a fixed exchange

# Pull Request - Checklist  

- [x] Initial Manual Tests Passed
- [x] Double check modified code and verify it with the feature/task requirements
- [x] Formate code
- [x] Look for code duplication
- [x] Clear naming for variables and methods

### CODE REVIEW

**Review by OmarHatem28 — CHANGES_REQUESTED** [review]
Please revert all the formatting changes, and only commit the actual changes

File: lib/core/amount_validator.dart [hunk 1]
```diff
@@ -3,17 +3,43 @@ import 'package:cake_wallet/generated/i18n.dart';
...
   }
 
   late final SymbolsAmountValidator symbolsAmountValidator;
 
   late final DecimalAmountValidator decimalAmountValidator;
 
-  String? call(String? value) => symbolsAmountValidator(value) ?? decimalAmountValidator(value);
+   late final AmountMinValidator amountMinValidator;
+
+   late final AmountMaxValidator amountMaxValidator;
```
Comment (OmarHatem28) [inline lib/core/amount_validator.dart:29]: indentation!

File: lib/src/screens/exchange/exchange_page.dart [hunk 2]
```diff
@@ -456,9 +456,13 @@ class ExchangePage extends BasePage {
...
-          exchangeViewModel.changeDepositAmount(
-              amount: depositAmountController.text);
-          exchangeViewModel.isReceiveAmountEntered = false;
+      if (exchangeViewModel.checkIfInputMeetsMinOrMaxCondition(
+                   depositAmountController.text) ==
+               true){
+  exchangeViewModel.changeDepositAmount(
+      amount: depositAmountController.text);
+  exchangeViewModel.isReceiveAmountEntered = false;
+}
```
Comment (OmarHatem28) [inline lib/src/screens/exchange/exchange_page.dart:465]: - format
- duplicate boolean check (i.e bool == true)

File: lib/src/screens/exchange/exchange_page.dart [hunk 3]
```diff
@@ -469,9 +473,13 @@ class ExchangePage extends BasePage {
...
-          exchangeViewModel.changeReceiveAmount(
-              amount: receiveAmountController.text);
-          exchangeViewModel.isReceiveAmountEntered = true;
+         if (exchangeViewModel.checkIfInputMeetsMinOrMaxCondition(
+                   receiveAmountController.text) ==
+               true) {
+  exchangeViewModel.changeReceiveAmount(
+      amount: receiveAmountController.text);
+  exchangeViewModel.isReceiveAmountEntered = true;
+}
```
Comment (OmarHatem28) [inline lib/src/screens/exchange/exchange_page.dart:482]: - format
- duplicate boolean check (i.e bool == true)

File: lib/view_model/exchange/exchange_view_model.dart [hunk 4]
```diff
@@ -305,7 +308,19 @@ abstract class ExchangeViewModelBase with Store {
...
+
+     if (_enteredAmount < minLimit) {
+       depositAmount = '';
+       return;
+     }
+
+     if (_enteredAmount > maxLimit && isFixedRateMode) {
+       depositAmount = '';
+       return;
+     }
```
Comment (OmarHatem28) [inline lib/view_model/exchange/exchange_view_model.dart:323]: - indentation
- if the `limits.max` is null, then it should be uncapped, you are resetting it to zero so the condition will always be true.
- also why are you checking on whether its a fixed rate or not with the maxlimit?

File: lib/view_model/exchange/exchange_view_model.dart [hunk 5]
```diff
@@ -320,6 +335,23 @@ abstract class ExchangeViewModelBase with Store {
...
+     if (_enteredAmount < minLimit) {
+       return false;
+     }
+
+     if (_enteredAmount > maxLimit && isFixedRateMode) {
+       return false;
+     }
+
+     return true;
+   }
```
Comment (OmarHatem28) [inline lib/view_model/exchange/exchange_view_model.dart:352]: - indentation
- if the limits.max is null, then it should be uncapped, you are resetting it to zero so the condition will always be true.
- also why are you checking on whether its a fixed rate or not with the maxlimit?

File: lib/core/amount_validator.dart [hunk 6]
```diff
@@ -49,6 +75,75 @@ class DecimalAmountValidator extends TextValidator {
...
+     return valueInDouble > minInDouble;
+   }
+
+   double? parseToDouble(String value) {
+     final data = double.tryParse(value.replaceAll(',', '.'));
+     return data;
+   }
+ }
+
+ class AmountMaxValidator extends Validator<String> {
```
Comment (OmarHatem28) [inline lib/core/amount_validator.dart:113]: indentation

**Review by OmarHatem28 — CHANGES_REQUESTED** [review]
will continue the review after the indentations and formatting is resolved, please double-check the rest of the added code for invalid indentation, extra empty lines (between different parts of code only 1 empty line is enough, translation files (`fr`)), and the comments I added

File: lib/core/amount_validator.dart (hunk 1 above)
Comment (Blazebrain) [reply lib/core/amount_validator.dart]: Fixed

File: cw_monero/example/pubspec.lock
[the remark this answers is not in the document]
Comment (Blazebrain) [reply cw_monero/example/pubspec.lock:1]: Fixed it

File: lib/src/screens/exchange/exchange_page.dart (hunk 2 above)
Comment (Blazebrain) [reply lib/src/screens/exchange/exchange_page.dart]: Remove the duplicated boolean and fixed the formatting

File: lib/src/screens/exchange/exchange_page.dart (hunk 3 above)
Comment (Blazebrain) [reply lib/src/screens/exchange/exchange_page.dart]: Removed the double boolean and fixed the formatting

File: lib/view_model/exchange/exchange_view_model.dart (hunk 4 above)
Comment (Blazebrain) [reply lib/view_model/exchange/exchange_view_model.dart]: Fixed the indentation, made it uncapped when the `limits.max` is null and removed the check

File: lib/view_model/exchange/exchange_view_model.dart (hunk 5 above)
Comment (Blazebrain) [reply lib/view_model/exchange/exchange_view_model.dart]: Fixed the indentation, made it uncapped when the `limits.max` is null and removed the check

File: lib/core/amount_validator.dart (hunk 6 above)
Comment (Blazebrain) [reply lib/core/amount_validator.dart]: Fixed the indentation issues

**Review by OmarHatem28 — APPROVED** [review]

**Pull request merged**